internal/proxy/cache.go

```diff
@@ -0,0 +1,199 @@
+package proxy
+
+import (
+	"bytes"
+	"context"
+	"encoding/hex"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"log"
+	"net/http"
+	"strings"
+	"time"
+
+	"github.com/hashicorp/vault-lambda-extension/internal/config"
+	"github.com/hashicorp/vault/api"
+	"github.com/hashicorp/vault/sdk/helper/cryptoutil"
+	"github.com/hashicorp/vault/sdk/helper/strutil"
+	gocache "github.com/patrickmn/go-cache"
+)
+
+const (
+	// Return a cached response if it exists, otherwise fall back to Vault and
+	// cache the response
+	headerOptionCacheable = "cache"
+
+	// Send the request to Vault and cache the response
+	headerOptionRecache = "recache"
+
+	// Ignore the cache and send the request to Vault, do not cache the response
+	headerOptionNocache = "nocache"
```

Comment:
The [standard cache-control headers](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Cache-Control) use a slightly different directive for these two behavior. `no-cache` is equivalent to `recache` and `no-store` is equivalent to no cache. These are vault-specific cache headers though, so I'll leave it up to you on whether we want to use the same naming convention.

Reply:
That's a good point, though I do think the headers as-is are clearer as to their meaning than those of the standard. However I don't like that there is overlap with the standard headers, since that could lead to confusion, so I'm happy to change them if we just want to stick with the standard.

Reply:
That's fair, these are not strictly the same standard cache-control headers.